src/PreparedWrites.h

```diff
@@ -0,0 +1,227 @@
+               uint64_t version, uint32_t timestamp,
+               Buffer& buffer, uint32_t *length = NULL);
+    PreparedOp(Buffer& buffer, uint32_t offset, uint32_t length);
+
+    class Header {
+      public:
+        Header(WireFormat::TxPrepare::OpType type,
+               uint64_t clientId,
+               uint64_t rpcId,
+               uint32_t participantCount)
+            : type(type),
+              clientId(clientId),
+              rpcId(rpcId),
+              participantCount(participantCount),
+              checksum(0)
+        {
+        }
+
+        /// Type of the staged operation.
+        WireFormat::TxPrepare::OpType type;
+
+        /// leaseId given for this prepare.
+        uint64_t clientId;
+
+        /// rpcId given for this prepare.
+        uint64_t rpcId;
+
+        /// Number of objects participating current transaction.
+        uint32_t participantCount;
+
+        /// CRC32C checksum covering everything but this field, including the
+        /// keys and the value.
+        uint32_t checksum;
+
+    } __attribute__((__packed__));
+    static_assert(sizeof(Header) == 28,
+        "Unexpected serialized PreparedOp size");
+
+    /// Copy of the PreparedOp header.
+    Header header;
+
+    /// Pointer to the array of #WireFormat::TxParticipant which contains
+    /// all information of objects participating transaction.
+    WireFormat::TxParticipant* participants;
+
+    /// Object to be written during COMMIT phase.
+    /// Its value is empty if OpType is READ or REMOVE; it only contains
+    /// key information in that case.
+    Object object;
+
+    void assembleForLog(Buffer& buffer);
+    bool checkIntegrity();
+    uint32_t computeChecksum();
+
+    DISALLOW_COPY_AND_ASSIGN(PreparedOp);
+};
+
+class PreparedOpTombstone {
+  public:
+    PreparedOpTombstone(PreparedOp& op, uint64_t segmentId);
+    explicit PreparedOpTombstone(Buffer& buffer, uint32_t offset = 0);
+
+    void assembleForLog(Buffer& buffer);
+    bool checkIntegrity();
+    uint32_t computeChecksum();
+
+    /**
+     * This data structure defines the format of an preparedOp's tombstone
+     * in a master server's log.
+     */
+    class Header {
+      public:
+        Header(uint64_t tableId,
+               KeyHash keyHash,
+               uint64_t leaseId,
+               uint64_t rpcId,
+               uint64_t segmentId)
+            : tableId(tableId),
+              keyHash(keyHash),
+              leaseId(leaseId),
+              rpcId(rpcId),
+              segmentId(segmentId),
+              checksum(0)
+        {
+        }
+
+        /// TableId for log distribution during recovery.
+        uint64_t tableId;
+
+        /// KeyHash for log distribution during recovery.
+        KeyHash keyHash;
+
+        /// leaseId given for the prepare that this tombstone is for.
+        /// A (leaseId, rpcId) tuple uniquely identifies a preparedOp log entry.
+        uint64_t leaseId;
```

Comment:
leaseId => clientLeaseId

Also, mentioned in the comment that this is the lease id for the client that initiated the transaction.